source/extensions/quic_listeners/quiche/envoy_quic_packet_writer.cc

```diff
@@ -0,0 +1,51 @@
+#include "extensions/quic_listeners/quiche/envoy_quic_packet_writer.h"
+
+#pragma GCC diagnostic push
+
+// QUICHE allows unused parameters.
+#pragma GCC diagnostic ignored "-Wunused-parameter"
+// QUICHE uses offsetof().
+#pragma GCC diagnostic ignored "-Winvalid-offsetof"
+
+#include "quiche/quic/core/quic_types.h"
+
+#pragma GCC diagnostic pop
+
+#include "extensions/quic_listeners/quiche/envoy_quic_utils.h"
+#include "common/buffer/buffer_impl.h"
+
+namespace Envoy {
+namespace Quic {
+
+quic::WriteResult EnvoyQuicPacketWriter::WritePacket(const char* buffer, size_t buf_len,
+                                                     const quic::QuicIpAddress& self_ip,
+                                                     const quic::QuicSocketAddress& peer_address,
+                                                     quic::PerPacketOptions* options) {
+  ASSERT(options == nullptr, "Per packet option is not supported yet.");
+  ASSERT(!write_blocked_, "Cannot write while IO handle is blocked.");
```

Comment:
I think it shouldn't be a ASSERT when `write_blocked_` is true. For the same reason that calling ::sendmsg on a blocked socket will not crash the program. If write_blocked_ is true, we can either return `WRITE_STATUS_BLOCKED` directly, or just forcefully call `listener_.send`.

Reply:
sure, I'm ok as long as it only crashes in debug build.